testutil/fuse/fuse.go

```diff
@@ -0,0 +1,238 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fuse
+
+import (
+	"fmt"
+	"github.com/prometheus/tsdb/testutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/hanwen/go-fuse/fuse"
+	"github.com/hanwen/go-fuse/fuse/nodefs"
+	"github.com/hanwen/go-fuse/fuse/pathfs"
+)
+
+type HookFs struct {
+	Original   string
+	Mountpoint string
+	FsName     string
+	fs         pathfs.FileSystem
+	hook       Hook
+}
+
+func NewHookFs(original string, mountpoint string, hook Hook) (*HookFs, error) {
+	loopbackfs := pathfs.NewLoopbackFileSystem(original)
+	hookfs := &HookFs{
+		Original:   original,
+		Mountpoint: mountpoint,
+		FsName:     "hookfs",
+		fs:         loopbackfs,
+		hook:       hook,
+	}
+	return hookfs, nil
+}
+
+func (h *HookFs) String() string {
+	return fmt.Sprintf("HookFs{Original=%s, Mountpoint=%s, FsName=%s, Underlying fs=%s, hook=%s}",
+		h.Original, h.Mountpoint, h.FsName, h.fs.String(), h.hook)
+}
+
+func (h *HookFs) SetDebug(debug bool) {
+	h.fs.SetDebug(debug)
+}
+
+func (h *HookFs) GetAttr(name string, context *fuse.Context) (*fuse.Attr, fuse.Status) {
+	return h.fs.GetAttr(name, context)
+}
+
+func (h *HookFs) Chmod(name string, mode uint32, context *fuse.Context) fuse.Status {
+	return h.fs.Chmod(name, mode, context)
+}
+
+func (h *HookFs) Chown(name string, uid uint32, gid uint32, context *fuse.Context) fuse.Status {
+	return h.fs.Chown(name, uid, gid, context)
+}
+
+func (h *HookFs) Utimens(name string, Atime *time.Time, Mtime *time.Time, context *fuse.Context) fuse.Status {
+	return h.fs.Utimens(name, Atime, Mtime, context)
+}
+
+func (h *HookFs) Truncate(name string, size uint64, context *fuse.Context) fuse.Status {
+	return h.fs.Truncate(name, size, context)
+}
+
+func (h *HookFs) Access(name string, mode uint32, context *fuse.Context) fuse.Status {
+	return h.fs.Access(name, mode, context)
+}
+
+func (h *HookFs) Link(oldName string, newName string, context *fuse.Context) fuse.Status {
+	return h.fs.Link(oldName, newName, context)
+}
+
+func (h *HookFs) Mkdir(name string, mode uint32, context *fuse.Context) fuse.Status {
+	return h.fs.Mkdir(name, mode, context)
+}
+
+func (h *HookFs) Mknod(name string, mode uint32, dev uint32, context *fuse.Context) fuse.Status {
+	return h.fs.Mknod(name, mode, dev, context)
+}
+
+func (h *HookFs) Rename(oldName string, newName string, context *fuse.Context) fuse.Status {
+	preHooked, err := h.hook.PreRename(oldName, newName)
+	if preHooked {
+		if err != nil {
+
+			return fuse.ToStatus(err)
+		}
+	}
+
+	status := h.fs.Rename(oldName, newName, context)
```

Comment:
after the struct embedding this will become:
```suggestion
	status := h.LoopBackFs.Rename(oldName, newName, context)
```